File: server/game-engine/src/main/java/com/deepdame/engine/core/logic/GameEngine.java
package com.deepdame.engine.core.logic;

import com.deepdame.engine.core.model.*;

import java.util.List;

public class GameEngine {

    private final MoveValidator validator;

    public GameEngine() {
        this.validator = new MoveValidator();
    }

    public GameState applyMove(GameState state, Move move){
        Board board = state.getBoard();
        PieceType player = state.getCurrentTurn();

        if (!validator.isValidMove(board, move, player)){
            throw new IllegalArgumentException("Invalide move");
        }

        Piece piece =  board.getPiece(move.from());
        board.removePiece(move.from());
        board.setPiece(move.to(), piece);

        if (move.isJump()){
            Position jumpedPos = move.getJumpedPosition();
            board.removePiece(jumpedPos);
        }

        checkPromotion(board, piece, move.to());
        checkGameOver(state);

        if (!state.isGameOver()){
            state.switchTurn();
        }

        return state;
    }

    public List<Move> getLegalMoves(Board board, PieceType player){
        return validator.getLegalMoves(board, player);
    }

    private void checkPromotion(Board board ,Piece piece, Position position){
        if (piece.type() == PieceType.BLACK && position.row() == 0){
            Piece kingPiece = board.getPiece(position).promote();
            board.setPiece(position, kingPiece);
        }else if (piece.type() == PieceType.WHITE && position.row() == 7){
            Piece kingPiece = board.getPiece(position).promote();
            board.setPiece(position, kingPiece);
        }
    }

    private void checkGameOver(GameState state){
        PieceType nextTurn = (state.getCurrentTurn() == PieceType.BLACK) ? PieceType.WHITE : PieceType.BLACK;
        List<Move> nextMoves = validator.getLegalMoves(state.getBoard(), nextTurn);

        if (nextMoves.isEmpty()){
            state.finishGame(state.getCurrentTurn());
        }
    }
}

File: client/deepdame/lib/game-engine/logic/game_engine.dart
import '../model/board.dart';
import '../model/game_state.dart';
import '../model/move.dart';
import '../model/piece.dart';
import '../model/piece_type.dart';
import '../model/position.dart';
import 'move_validator.dart';

class GameEngine {
  final MoveValidator validator;

  GameEngine() : validator = MoveValidator();

  GameState applyMove(GameState state, Move move) {
    Board board = state.board;
    PieceType player = state.currentTurn;

    if (!validator.isValidMove(board, move, player)) {
      throw ArgumentError("Invalid move");
    }

    Piece piece = board.getPiece(move.from)!;
    
    board.removePiece(move.from);
    board.setPiece(move.to, piece);

    if (move.isJump) {
      Position? jumpedPos = move.jumpedPosition;
      if (jumpedPos != null) {
        board.removePiece(jumpedPos);
      }
    }

    _checkPromotion(board, piece, move.to);
    _checkGameOver(state);

    if (!state.isGameOver) {
      state.switchTurn();
    }

    return state;
  }

  List<Move> getLegalMoves(Board board, PieceType player) {
    return validator.getLegalMoves(board, player);
  }

  void _checkPromotion(Board board, Piece piece, Position position) {
    
    if (piece.type == PieceType.black && position.row == 0) {
      Piece kingPiece = board.getPiece(position)!.promote();
      board.setPiece(position, kingPiece);
    } 
    else if (piece.type == PieceType.white && position.row == Board.size - 1) {
      Piece kingPiece = board.getPiece(position)!.promote();
      board.setPiece(position, kingPiece);
    }
  }

  void _checkGameOver(GameState state) {
    PieceType nextTurn = (state.currentTurn == PieceType.black) 
        ? PieceType.white 
        : PieceType.black;
    
    List<Move> nextMoves = validator.getLegalMoves(state.board, nextTurn);

    if (nextMoves.isEmpty) {
      state.finishGame(state.currentTurn);
    }
  }
}